include/revng/BasicAnalyses/GeneratedCodeBasicInfo.h

```diff
@@ -250,41 +243,77 @@ class GeneratedCodeBasicInfo : public llvm::ModulePass {
 
   /// \brief Return the program counter of the next (i.e., fallthrough)
   ///        instruction of \p TheInstruction
-  MetaAddress getNextPC(llvm::Instruction *TheInstruction) const {
+  MetaAddress getNextPC(llvm::Instruction *TheInstruction) {
     auto Pair = getPC(TheInstruction);
     return Pair.first + Pair.second;
   }
 
-  llvm::CallInst *getFunctionCall(llvm::BasicBlock *BB) const {
+  static llvm::CallInst *getFunctionCall(llvm::BasicBlock *BB) {
     return getFunctionCall(BB->getTerminator());
   }
 
   // TODO: is this a duplication of FunctionCallIdentification::isCall?
   // TODO: we could unpack the information too
-  llvm::CallInst *getFunctionCall(llvm::Instruction *T) const {
+  static llvm::CallInst *getFunctionCall(llvm::Instruction *T) {
+    revng_assert(T != nullptr);
     revng_assert(T->isTerminator());
-    auto It = T->getIterator();
-    auto End = T->getParent()->begin();
-    while (It != End) {
-      It--;
-      if (llvm::CallInst *Call = getCallTo(&*It, "function_call"))
+    llvm::Instruction *Previous = getPrevious(T);
+    while (Previous != nullptr && isMarker(Previous)) {
+      if (auto *Call = getCallTo(Previous, "function_call"))
         return Call;
 
-      if (not isMarker(&*It))
-        return nullptr;
+      Previous = getPrevious(Previous);
     }
 
     return nullptr;
   }
```

Comment:
`getPrevious` might be nicer but it's less efficient.
Did you find the function so unreadable?

Reply:
I didn't, but I remember talking with @pfez about these lines, who also suggested using `isMarker` and `getPrevious` (if I recollect correctly). Looks cleaner to me this way, if it's not too less efficient, but we can keep only `isMarker` if you say.

Reply:
Uhm, OK, make sure that this improvement gets its own commit.

Reply:
My point was that the previous version did not work if the argument was the first instruction in a basic block, because it would never have entered the `while (It != End)`. Using `getPrevious()` works as expected all the times.

Reply:
You also have `Instruction * Instruction::getPrevNode()`, which is defined by llvm, and not by ourselves, and does exactly the same thing. Maybe we should avoid defining our own helpers that are identical to llvm stuff that is already available.

Reply:
I'm not sure why this was changed, the original version looked fine to me.
However, I'd suggest not to make too many unrelated changes to IRHelpers.h in an opportunistic way.
We need to sit down and review in a single shot `revng.h`, `IRHelpers.h`, `GCBI.h` and reorganize them. But this is not the right time.

Reply:
Noted down to fix in the future.